### PR TITLE
WIP Plugins encryption key rotation

### LINKED ISSUE #295: Java: Plug in for message encryption - encrypt
**Problem**
No plugin exists

**Solution**
Implement a plugin to encrypt a message using some encryption method

### LINKED ISSUE #342: Java-core: Change encryption from AES/CBC to AES/GCM
**Detailed Description**

**Actual Behavior**

**Expected Behavior**

**Environment:**
- Java: [e.g. OpenJDK 11.0.4]
- Node: [e.g. v12.9.1-x86]
- OS: [e.g. Ubuntu 18.04]
- Version: [e.g. 1.0.1]

**Additional Context**
Add any other context about the problem here. Attach any logs here, if applicable.

### PR DESCRIPTION
**Pull request by zoperado**
**Detailed description**:

**Which issue(s) this PR fixes**:
Fixes #342 
Fixes #295  

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [ ] Changelog updated
- [ ] Readme updated
- [ ] Tests updated

### CODE REVIEW

File: hcs-sxc-java-core/src/main/java/com/hedera/hcs/sxc/callback/OnHCSMessageCallback.java [hunk 1]
```diff
@@ -53,30 +78,59 @@ public OnHCSMessageCallback (HCSCore hcsCore) throws Exception {
...
                 pushUntilCompleteMessage(messagePart, this.hcsCore.getMessagePersistence());
 
-        if (messageEnvelopeOptional.isPresent()){
+        if (messageEnvelopeOptional.isPresent()){ // is present if all parts received
             this.hcsCore.getMessagePersistence().storeApplicationMessage(messageEnvelopeOptional.get().getApplicationMessageId(), messageEnvelopeOptional.get());
-            notifyObservers( messageEnvelopeOptional.get().toByteArray(), messageEnvelopeOptional.get().getApplicationMessageId());
+                
+            ApplicationMessage appMessage = messageEnvelopeOptional.get();
+            
+            if(this.encryptMessages){
```
Comment (gregscullard) [inline hcs-sxc-java-core/src/main/java/com/hedera/hcs/sxc/callback/OnHCSMessageCallback.java]: This flag is to indicate messages should be encrypted, there is an override on message encryption (this boolean could be false and a message be encrypted, the opposite is true too), we should have a flag in the application message that indicates whether the message was encrypted or not by the sender and use that here to determine if decryption is necessary (maybe raise a separate issue and assume appnet-wide encryption for now)

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml [hunk 2]
```diff
@@ -27,6 +27,19 @@ services:
       - 8082:8082
 
   Carlos:
+<<<<<<< HEAD
```
Comment (gregscullard) [inline hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml]: Merge gone wrong ?

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml [hunk 3]
```diff
@@ -27,6 +27,19 @@ services:
...
+    restart: on-failure
+    environment:
+      APP_ID: 2
+      PUBKEY: 302a300506032b6570032100bcabbb31c4c6418ea323e02dd46c060f82936141b5d2d2a1da89e59e1267ab6c
+      ROLES: "BUYER,SELLER"
+      PAYMENT_ACCOUNT_DETAILS: "22-11-14 70925098"
+      COLOR: 275f4f
+    ports:
+      - 8083:8083
+=======
```
Comment (gregscullard) [inline hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml]: Merge gone wrong ?

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml [hunk 4]
```diff
@@ -38,6 +51,7 @@ services:
      COLOR: 275f4f
    ports:
      - 8083:8083
+>>>>>>> 3eb863335fb668ddbf1ae6e9f9e0b7a1261d4483
```
Comment (gregscullard) [inline hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml]: Merge gone wrong ?

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml [hunk 5]
```diff
@@ -53,6 +67,45 @@ services:
       - 8084:8084
 
   Erica:
+<<<<<<< HEAD
```
Comment (gregscullard) [inline hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml]: Merge gone wrong ?

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml [hunk 6]
```diff
@@ -53,6 +67,45 @@ services:
...
+    restart: on-failure
+    environment:
+      APP_ID: 6
+      PUBKEY: 302a300506032b6570032100bcabbb31c4c6418ea323e02dd46c060f82936141b5d2d2a1da89e59e1267ab7a
+      ROLES: "DEMO"
+      PAYMENT_ACCOUNT_DETAILS: "10-32-52 55424452"
+      COLOR: 4f0500
+    ports:
+      - 8087:8087
+=======
```
Comment (gregscullard) [inline hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml]: Merge gone wrong ?

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml [hunk 7]
```diff
@@ -90,6 +143,7 @@ services:
      COLOR: 4f0500
    ports:
      - 8087:8087
+>>>>>>> 3eb863335fb668ddbf1ae6e9f9e0b7a1261d4483
```
Comment (gregscullard) [inline hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml]: Merge gone wrong ?

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/src/main/java/com/hedera/hcsapp/controllers/CreditsController.java [hunk 8]
```diff
@@ -173,7 +173,7 @@ public CreditsController() throws Exception {
             
             credit = creditRepository.save(credit);
             new OutboundHCSMessage(appData.getHCSCore())
-                  .overrideEncryptedMessages(false)
+                  .overrideEncryptedMessages(true)
```
Comment (gregscullard) [inline hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/src/main/java/com/hedera/hcsapp/controllers/CreditsController.java]: See other comment wrt override, think we need to raise a separate issue on this.

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/src/main/java/com/hedera/hcsapp/integration/HCSIntegration.java [hunk 9]
```diff
@@ -244,8 +245,10 @@ public void processHCSMessage(HCSResponse hcsResponse) {
         
                                 settlementRepository.save(settlement);
         
-                                new OutboundHCSMessage(appData.getHCSCore()).overrideEncryptedMessages(false)
-                                        .overrideMessageSignature(false).sendMessage(appData.getTopicIndex(), newSettlementBPM.toByteArray());
+                                OutboundHCSMessage o =  new OutboundHCSMessage(appData.getHCSCore());
+                                        o.overrideEncryptedMessages(true);
```
Comment (gregscullard) [inline hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/src/main/java/com/hedera/hcsapp/integration/HCSIntegration.java]: Need issue raised here (see above)

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/src/main/java/com/hedera/hcsapp/integration/HCSIntegration.java [hunk 10]
```diff
@@ -244,8 +245,10 @@ public void processHCSMessage(HCSResponse hcsResponse) {
...
+                                        o.overrideMessageSignature(false);
```
Comment (gregscullard) [inline hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/src/main/java/com/hedera/hcsapp/integration/HCSIntegration.java]: Message signature should still be optional on unencrypted messages, but HAS to be included on encrypted messages.

File: hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/main/java/com/hedera/hcs/sxc/plugin/cryptography/cryptography/Cryptography.java [hunk 11]
```diff
@@ -1,5 +1,6 @@
-package com.hedera.hcs.sxc.cryptography;
-
+package com.hedera.hcs.sxc.plugin.cryptography.cryptography;
```
Comment (gregscullard) [inline hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/main/java/com/hedera/hcs/sxc/plugin/cryptography/cryptography/Cryptography.java]: Probably should name plug-in to match cryptography used here so others can be added later. Cryptography is too generic.

File: hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/main/java/com/hedera/hcs/sxc/plugin/cryptography/keyrotation/KeyRotation.java [hunk 12]
```diff
@@ -1,4 +1,4 @@
-package com.hedera.hcs.sxc.cryptography;
+package com.hedera.hcs.sxc.plugin.cryptography.keyrotation;
```
Comment (gregscullard) [inline hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/main/java/com/hedera/hcs/sxc/plugin/cryptography/keyrotation/KeyRotation.java]: Plugin name should indicate which key rotation scheme is being used to allow for others in the future.

File: hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/test/java/com/hedera/hcs/sxc/plugin/cryptography/cryptography/CryptographyTest.java [hunk 13]
```diff
@@ -1,53 +1,51 @@
...
+    String privateKeyHexEncoded = null;
+    byte[] publicKeyBytes = null;
+    String publicKeyKexEncoded = null;
+    
+    public CryptographyTest() {
+    }
+
+    @BeforeAll
+    public static void initClass(){
+        try {
```
Comment (gregscullard) [inline hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/test/java/com/hedera/hcs/sxc/plugin/cryptography/cryptography/CryptographyTest.java]: prefix method with `test` > suggest rename to testInitClass()

File: hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/test/java/com/hedera/hcs/sxc/plugin/cryptography/cryptography/CryptographyTest.java [hunk 14]
```diff
@@ -1,53 +1,51 @@
...
+            //System.out.println(StringUtils.byteArrayToHexString(secretKey));
+        } catch (Exception ex) {
+            Logger.getLogger(CryptographyTest.class.getName()).log(Level.SEVERE, null, ex);
+        }
+    }
+    
+    
+    @Test
+    public void encryptAndDecrypt() throws Exception {
+        byte[] encrypt = Cryptography.load().encrypt(secretKey, StringUtils.stringToByteArray(cleartext));
```
Comment (gregscullard) [inline hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/test/java/com/hedera/hcs/sxc/plugin/cryptography/cryptography/CryptographyTest.java]: prefix method with `test` > suggest rename to testEncryptAndDecrypt()

File: hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/test/java/com/hedera/hcs/sxc/plugin/cryptography/keyrotation/KeyRotationTest.java [hunk 15]
```diff
@@ -1,12 +1,11 @@
-package com.hedera.hcs.sxc.cryptography;
+package com.hedera.hcs.sxc.plugin.cryptography.keyrotation;
```
Comment (gregscullard) [inline hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/test/java/com/hedera/hcs/sxc/plugin/cryptography/keyrotation/KeyRotationTest.java]: Ensure test method names start with `test`

File: hcs-sxc-java-core/src/main/java/com/hedera/hcs/sxc/callback/OnHCSMessageCallback.java (hunk 1 above)
Comment (gregscullard) [reply hcs-sxc-java-core/src/main/java/com/hedera/hcs/sxc/callback/OnHCSMessageCallback.java]: See issue #371

Comment (gregscullard) [reply hcs-sxc-java-core/src/main/java/com/hedera/hcs/sxc/callback/OnHCSMessageCallback.java]: See issue #371

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/src/main/java/com/hedera/hcsapp/controllers/CreditsController.java (hunk 8 above)
Comment (gregscullard) [reply hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/src/main/java/com/hedera/hcsapp/controllers/CreditsController.java]: See issue #371

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/src/main/java/com/hedera/hcsapp/integration/HCSIntegration.java (hunk 9 above)
Comment (gregscullard) [reply hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/src/main/java/com/hedera/hcsapp/integration/HCSIntegration.java]: See issue #371

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/src/main/java/com/hedera/hcsapp/integration/HCSIntegration.java (hunk 10 above)
Comment (gregscullard) [reply hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/src/main/java/com/hedera/hcsapp/integration/HCSIntegration.java]: See issue #373

File: hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/main/java/com/hedera/hcs/sxc/plugin/cryptography/keyrotation/KeyRotation.java (hunk 12 above)
Comment (gregscullard) [inline hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/main/java/com/hedera/hcs/sxc/plugin/cryptography/keyrotation/KeyRotation.java]: Can you rename instances of `alice` to `initiator` and `bob` to `responder` ?

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml (hunk 2 above)
Comment (zoperado) [reply hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml]: I had disabled the examples. merging bits only until things work.

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml (hunk 3 above)
Comment (zoperado) [reply hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml]: I had disabled the examples. merging bits only until things work.

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml (hunk 4 above)
Comment (zoperado) [reply hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml]: I had disabled the examples. merging bits only until things work.

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml (hunk 5 above)
Comment (zoperado) [reply hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml]: I had disabled the examples. merging bits only until things work.

File: hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml (hunk 6 above)
Comment (zoperado) [reply hcs-sxc-java-examples/hcs-sxc-java-settlement-demo/config/docker-compose.sample.yml]: I had disabled the examples. merging bits only until things work.

File: hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/test/java/com/hedera/hcs/sxc/plugin/cryptography/cryptography/CryptographyTest.java (hunk 13 above)
Comment (zoperado) [reply hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/test/java/com/hedera/hcs/sxc/plugin/cryptography/cryptography/CryptographyTest.java]: done

File: hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/test/java/com/hedera/hcs/sxc/plugin/cryptography/keyrotation/KeyRotationTest.java (hunk 15 above)
Comment (zoperado) [reply hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/test/java/com/hedera/hcs/sxc/plugin/cryptography/keyrotation/KeyRotationTest.java]: done

File: hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/main/java/com/hedera/hcs/sxc/plugin/cryptography/keyrotation/KeyRotation.java (hunk 12 above)
Comment (zoperado) [reply hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/main/java/com/hedera/hcs/sxc/plugin/cryptography/keyrotation/KeyRotation.java]: done

File: hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/test/java/com/hedera/hcs/sxc/plugin/cryptography/cryptography/CryptographyTest.java (hunk 14 above)
Comment (zoperado) [reply hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/test/java/com/hedera/hcs/sxc/plugin/cryptography/cryptography/CryptographyTest.java]: done

File: hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/main/java/com/hedera/hcs/sxc/plugin/cryptography/keyrotation/KeyRotation.java (hunk 12 above)
Comment (zoperado) [reply hcs-sxc-java-plugins/hcs-sxc-java-plugins-cryptography/src/main/java/com/hedera/hcs/sxc/plugin/cryptography/keyrotation/KeyRotation.java]: TODO, creating issue

**Pull request closed without merging**